aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,467 @@
+      "grad and output sparse dimensions must be equal");
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
+      // assuming that the type of input._indices() entries is int64_t
+      cpu_sparse_coo_softmax_backward<scalar_t, int64_t, false>(grad_input, grad, output, dim);
+  });
+  return grad_input;
+}
+
+Tensor log_softmax_backward_sparse_cpu(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim_,
+    const Tensor& input_) {
+  TensorArg grad_arg{grad_, "grad", 1}, output_arg{output_, "output", 2};
+  checkSameSize("log_softmax_backward", grad_arg, output_arg);
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+
+  auto grad = grad_.coalesce();
+  auto output = output_.coalesce();
+
+  Tensor grad_input = at::native::empty_like(output);
+  if (output.numel() == 0) {
+    return grad_input;
+  }
+  TORCH_CHECK(
+      dim >= 0 && dim < grad.dim(),
+      "dim must be non-negative and less than input dimensions");
+  TORCH_CHECK(
+              grad.sparse_dim() == output.sparse_dim(),
+      "grad and output sparse dimensions must be equal");
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
+      // assuming that the type of input._indices() entries is int64_t
+      cpu_sparse_coo_softmax_backward<scalar_t, int64_t, true>(grad_input, grad, output, dim);
+  });
+  return grad_input;
+}
+
+Tensor _sparse_softmax(const Tensor& input_, const int64_t dim_) {
+  auto result = [&]() {
+    NoNamesGuard guard;
+    return at::_sparse_softmax(input_, dim_, false);
+  }();
+  namedinference::propagate_names(result, input_);
+  return result;
+}
+
+Tensor _sparse_softmax(const Tensor& input_, const int64_t dim_, c10::optional<ScalarType> dtype) {
+  auto result = [&]() {
+    NoNamesGuard guard;
+    if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
+        return at::_sparse_softmax(input_, dim_, true);
+    } else {
+        Tensor converted = dtype.has_value() ? input_.toType(dtype.value()) : input_;
+        return at::_sparse_softmax(converted, dim_, false);
+    }
+  }();
+  namedinference::propagate_names(result, input_);
+  return result;
+}
+
+Tensor _sparse_softmax(const Tensor& self, Dimname dim, optional<ScalarType> dtype) {
+  return at::_sparse_softmax(self, dimname_to_position(self, dim), dtype);
+}
+
+Tensor _sparse_log_softmax(const Tensor& input_, const int64_t dim_) {
+  auto result = [&]() {
+    NoNamesGuard guard;
+    return at::_sparse_log_softmax(input_, dim_, false);
+  }();
+  namedinference::propagate_names(result, input_);
+  return result;
+}
+
+Tensor _sparse_log_softmax(const Tensor& input_, const int64_t dim_, c10::optional<ScalarType> dtype) {
+  auto result = [&]() {
+    NoNamesGuard guard;
+    if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
+        return at::_sparse_log_softmax(input_, dim_, true);
```

Comment:
Hmm, what's going on here?

Reply:
This is a shameless copy from aten/src/Aten/native/SoftMax.cpp. For explanation, when a user requests output dtype to be different from input dtype, by default, the input is converted to output dtype, and the softmax is applied to the converted input. However, for Half input & Float output case, the softmax cuda implementation keeps the input as it is (Half dtype) and the results are accumulated to (Float dtype) output. As I understand it, this approach allows saving CUDA device memory.